baseclasses/SolverHist.py

```diff
@@ -0,0 +1,210 @@
+    default, the history only stores the values from the latest iteration which are used for printing to the terminal.
+    Storing of the full solution history is enabled using the 'writeConvergenceHist' option
+    """
+
+    def __init__(self, options):
+        """
+        Create an empty history dictionary based on the current solver options
+        """
+
+        if options["writeconvergencehist"]:
+            length = int(options["maxiter"] + 1)
+            self.storeValues = True
+        else:
+            length = 1
+            self.storeValues = False
+
+        # Make a local copy of the options
+        self.options = options
+
+        # Define some defaults
+        self.defaults = {"values": {}, "valueFormat": {}}
+        self.defaults["valueFormat"]["float"] = "{:17.11e}"
+        self.defaults["values"]["float"] = np.full(length, np.nan)
+
+        self.defaults["valueFormat"]["int"] = "{:03d}"
+        self.defaults["values"]["int"] = [np.nan] * length
+
+        self.defaults["valueFormat"]["other"] = "{:^40}"
+        self.defaults["values"]["other"] = [None] * length
+
+        self.hist = {"values": {}, "valueFormat": {}, "nameFormat": {}, "varType": {}}
+        self.varsToPrint = []
+        self.latestIter = 0
+
+    def addField(self, Name, Type, ValueFormat=None, NameFormat=None, Print=False):
+        """Add a field to the solution history"""
+
+        if Print:
+            self.varsToPrint.append(Name)
+
+        if Type in [float, np.float64, np.float32]:
+            varType = "float"
+            testValue = 0.0
+        elif Type is int:
+            varType = "int"
+            testValue = 0
+        else:
+            varType = "other"
+            testValue = " "
+
+        self.hist["varType"][Name] = varType
+
+        if ValueFormat is None:
+            ValueFormat = copy.deepcopy(self.defaults["valueFormat"][varType])
+
+        if NameFormat is None:
+            testString = ValueFormat.format(testValue)
+            # If name is longer than the value then name format should be length of name and value format should be
+            # lengthened, otherwise make name format the same length as value format
+            if len(Name) > len(testString):
+                NameFormat = "{:^%i}" % len(Name)
+                spacesToAdd = len(Name) - len(testString)
+                ValueFormat = " " * (spacesToAdd - spacesToAdd // 2) + ValueFormat + " " * (spacesToAdd // 2)
+            else:
+                NameFormat = "{:^%i}" % len(testString)
+
+        # Now actually add the entry
+        self.hist["values"][Name] = copy.deepcopy(self.defaults["values"][varType])
+        self.hist["valueFormat"][Name] = ValueFormat
+        self.hist["nameFormat"][Name] = NameFormat
+
+    def addUserFuncVars(self, Solver, UserFunc):
+        """Add the variables generated by a user supplied function to those tracked by the SolverHist object"""
+
+        data = UserFunc(Solver)
```

Comment:
I am just a bit confused about the `UserFunc` signature here, we might want to add an example at some point, even if this method just takes it in.